components/st-microelectronics/stm-api.stanza

```diff
@@ -130,132 +138,177 @@ defn match-pin? (pin-ref:Ref power:String) :
   found-pins
 
-defn power-pins-list (stm:JITXObject vdd-pin:PowerPinNames) -> Tuple<JITXObject> :
-  find-pins-list(stm, to-string(vdd-pin))
-
 defn power-pins (stm:JITXObject) -> Tuple<JITXObject> :
   power-pins(stm, VDD)
 
+defn power-pins-list (stm:JITXObject vdd-pin:PwrPinNames|ReferencePinNames) -> Tuple<JITXObject> :
+  get-pin-list(stm, vdd-pin)
+
 defn gnd-pins (stm:JITXObject, gnd-pin:GndPinNames) -> Tuple<JITXObject> :
-  val found-pins = find-pins-list(stm, to-string(gnd-pin))
+  val found-pins = get-pin-list(stm, gnd-pin)
   if empty?(found-pins) :
-    fatal("%_ is missing an identifiable ground pin." % [ref(stm)])
+    fatal("%_ is missing an identifiable ground pin (%_)." % [ref(stm), to-string(found-pins[0])])
   found-pins
 
-defn gnd-pins-list (stm:JITXObject, gnd-pin:GndPinNames) -> Tuple<JITXObject> :
-  find-pins-list(stm, to-string(gnd-pin))
-
 defn gnd-pins (stm:JITXObject) -> Tuple<JITXObject> :
   gnd-pins(stm, VSS)
 
-public defn generate-named-bypass (stm:JITXObject, bypass-cap-package:Double|False, bypass-cap-pin:Double, vdd-net:PowerPinNames, gnd-net:GndPinNames) :
+defn gnd-pins-list (stm:JITXObject, gnd-pin:GndPinNames) -> Tuple<JITXObject> :
+  get-pin-list(stm, gnd-pin)
+
+public defn generate-pin-bypass (stm:JITXObject, vdd-net:PwrPinNames|ReferencePinNames, gnd-net:GndPinNames, cap-bypass-pin:Double) :
   inside pcb-module :
-    val power-pin-list = power-pins(stm, vdd-net)
-    val gnd-pin-list = gnd-pins(stm, gnd-net)
-    match(bypass-cap-package) :
-      (bp: Double) : bypass-cap-strap(power-pin-list[0], gnd-pin-list[0], bp)
-      (bp: False) : false
-    for p in power-pin-list do :
-      bypass-cap-strap(p,  gnd-pin-list[0], bypass-cap-pin)
-
-public defn generate-bypass (stm:JITXObject, bypass-package:Double, bypass-pin:Double) :
+    val power-pins = power-pins-list(stm, vdd-net)
+    val gnd-pins = gnd-pins-list(stm, gnd-net)
+    if not empty?(gnd-pins) :
+      for p in power-pins do :
+        bypass-cap-strap(p, gnd-pins[0], cap-bypass-pin)
+
+public defn generate-pin-bypass (stm:JITXObject, cap-bypass-pin:Double) :
+  generate-pin-bypass(stm, VDD, VSS, cap-bypass-pin)
+
+public defn generate-pin-bypass (stm:JITXObject) :
+  val settings = Settings(DEFAULT-SETTINGS)
+  generate-pin-bypass(stm, VDD, VSS, settings[`cap-bypass-pin])
+
+public defn generate-package-bypass (stm:JITXObject, vdd-net:PwrPinNames, gnd-net:GndPinNames, cap-bypass-package:Double) :
   inside pcb-module :
-    val power-pin-list = power-pins(stm, VDD)
-    val gnd-pin-list = gnd-pins(stm, VSS)
-    bypass-cap-strap(power-pin-list[0], gnd-pin-list[0], bypass-package)
-    for p in power-pin-list do :
-      bypass-cap-strap(p,  gnd-pin-list[0], bypass-pin)
+    val power-pins = power-pins-list(stm, vdd-net)
+    val gnd-pins = gnd-pins-list(stm, gnd-net)
+    if not empty?(power-pins) and not empty?(gnd-pins) :
+      bypass-cap-strap(power-pins[0], gnd-pins[0], cap-bypass-package)
```

Comment:
This is clean!
Should we inform the user if they requested a package bypass but one couldn't be created?

Reply:
I think that the user will have bigger issues if the bypass is not able to be created -- the connect-power(stm, "power", VDD, VSS) will fail hard. But I guess that more info is useful. Will add, thanks.

Reply:
Actually, the function will return false if no bypass-cap-strap is instantiated. We could always look at the return value for the function?

Reply:
Yes, the `false` is a good indicator already! Good point